README.md
# Swift Object Storage

## Installation
`npm install` or `yarn install`


## Example Usage
```javascript
const config = {
  objectStorageUrl: objectStorageUrl: 'https://dal05.objectstorage.softlayer.net/auth/v1.0/'
	user: 'username';
	key: 'key';

}
const client = new SwiftObjectStorage(config);

// Get remote file and store locally
client.getObjectWriteToStream(
  fs.createWriteStream('my-local-file.pdf'),
  'some-container'
  'remote-filepdf'
)
  .then() => console.log('done');
```

### Inject `onResponse` callback.

You can optionally pass an `onResponse` callback and it will be passed to Request's `.on('response')`. This is useful for when you need access to the response context

```javascript
client.getObjectWriteToStream(
  fs.createWriteStream('my-local-file.pdf'),
  'some-container'
  'remote-filepdf',
  {
    onResponse: (resp) => console.log('Response', resp);
  }
)
  .then() => console.log('done');
```


CHANGELOG.md
## [1.0.2](https://github.com/inkcloud/swift-object-storage/compare/v1.0.1...v1.0.2) (2018-08-02)


### Bug Fixes

* **SwiftObjectStorage:** Fix export ([dd7b656](https://github.com/inkcloud/swift-object-storage/commit/dd7b656))

## [1.0.1](https://github.com/inkcloud/swift-object-storage/compare/v1.0.0...v1.0.1) (2018-08-02)


### Bug Fixes

* **release:** Fixed publishing to NPM ([e145bdb](https://github.com/inkcloud/swift-object-storage/commit/e145bdb))

# 1.0.0 (2018-08-02)


### Bug Fixes

* **Docs:** Added generated documentation ([3c50b6d](https://github.com/inkcloud/swift-object-storage/commit/3c50b6d))


src/SwiftObjectStorage.ts
import request, { Response, Request } from 'request';
import Cache from 'node-cache';
import pump from 'pump';
import { WriteStream, ReadStream } from 'fs';

const SWIFT_TOKEN_KEY = 'SWIFT_TOKEN_KEY';
const DEFAULT_EXP_SECONDS = 60 * 20;

export interface TokenData {
  authToken: string | string[];
  objectPath: string;
}

export interface GetObjectAsStreamOptions {
  onResponse?: (resp: Response) => void;
}

export interface CreateObjectFromStreamOptions {
  onResponse?: (resp: Response) => void;
  expiresAfterSeconds?: number;
  headers?: {
    [key: string]: string;
  };
}

export interface SwiftObjectStorageOptions {
  objectStorageUrl: string;
  user: string;
  key: string;
  cacheExpireAfterSeconds?: number;
}

/**
 * Simple Swift Object Storage client
 */
export class SwiftObjectStorage {
  tokenCache: Cache;
  constructor(private options: SwiftObjectStorageOptions) {
    this.tokenCache = new Cache();
  }

  /**
   * Handles token authentication. Keeps the token data  cached so that
   * a new token doesn't have to be generated on every request
   */
  async getTokenData(): Promise<TokenData> {
    const tokenData: TokenData = this.tokenCache.get<TokenData>(SWIFT_TOKEN_KEY);

    if (tokenData) {
      return Promise.resolve(tokenData);
    }

    return new Promise<TokenData>((resolve, reject) => {
      request.get(
        {
          url: this.options.objectStorageUrl,
          headers: {
            'X-Auth-Key': this.options.key,
            'X-Auth-User': this.options.user
          }
        },
        (err, resp) => {
          if (resp.statusCode !== 200) {
            return reject(
              Error(
                `Error generating token: ${resp.statusCode} - ${
                  resp.statusMessage
                }`
              )
            );
          }

          const data = JSON.parse(resp.body);
          const newTokenData: TokenData = {
            objectPath: data.storage.public,
            authToken: resp.headers['x-auth-token']
          };

          const expAfterSeconds =
            this.options.cacheExpireAfterSeconds || DEFAULT_EXP_SECONDS;
          this.tokenCache.set(
            SWIFT_TOKEN_KEY,
            newTokenData,
            this.options.cacheExpireAfterSeconds
          );
          return resolve(newTokenData);
        }
      );
    });
  }

  /**
   * Returns a Request readable stream
   * @param container
   * @param filename
   */
  async getRawGetStream(container: string, filename: string): Promise<Request> {
    const tokenData = await this.getTokenData();
    return Promise.resolve(
      request.get({
        url: `${tokenData.objectPath}/${container}/${filename}`,
        headers: { 'X-Auth-Token': tokenData.authToken }
      })
    );
  }

  async getObjectWriteToStream(
    writeStream: WriteStream,
    container: string,
    filename: string,
    options: GetObjectAsStreamOptions = {}
  ) {
    const sourceStream = await this.getRawGetStream(container, filename);
    return new Promise((resolve, reject) => {
      sourceStream.on('response', resp => {
        if (Math.floor(resp.statusCode / 100) !== 2) {
          reject(
            Error(
              `Error generating token: ${resp.statusCode} - ${
                resp.statusMessage
              }`
            )
          );
          return;
        }

        if (options.onResponse) {
          options.onResponse(resp);
        }
      });

      pump(sourceStream, writeStream, err => {
        if (err) {
          reject(err);
          return;
        }

        resolve();
      });
    });
  }

  /**
   * Returns a Request writeable stream
   * @param container
   * @param filename
   * @param options
   */
  async getRawCreateStream(
    container,
    filename: string,
    options: CreateObjectFromStreamOptions = {}
  ): Promise<Request> {
    const tokenData = await this.getTokenData();
    return Promise.resolve(
      request.put({
        url: `${tokenData.objectPath}/${container}/${filename}`,
        timeout: 1000 * 60 * 10,
        headers: {
          ...options.headers,
          'X-Delete-After': options.expiresAfterSeconds,
          'X-Auth-Token': tokenData.authToken
        }
      })
    );
  }

  createObjectFromStream(
    sourceStream: ReadStream,
    container: string,
    filename: string,
    options: CreateObjectFromStreamOptions = {}
  ) {
    return new Promise(async (resolve, reject) => {
      const putStream = await this.getRawCreateStream(
        container,
        filename,
        options
      );
      putStream.on('response', resp => {
        if (Math.floor(resp.statusCode / 100) !== 2) {
          reject(
            Error(
              `Error generating token: ${resp.statusCode} - ${
                resp.statusMessage
              }`
            )
          );
          return;
        }

        if (options.onResponse) {
          options.onResponse(resp);
        }
      });

      pump(sourceStream, putStream, err => {
        if (err) {
          reject(err);
          return;
        }

        resolve();
      });
    });
  }
}


lib/SwiftObjectStorage.d.ts
/// <reference types="node" />
import { Response, Request } from "request";
import Cache from "node-cache";
import { WriteStream, ReadStream } from "fs";
export interface TokenData {
    authToken: string | string[];
    objectPath: string;
}
export interface GetObjectAsStreamOptions {
    onResponse?: (resp: Response) => void;
}
export interface CreateObjectFromStreamOptions {
    onResponse?: (resp: Response) => void;
    expiresAfterSeconds?: number;
    headers?: {
        [key: string]: string;
    };
}
export interface SwiftObjectStorageOptions {
    objectStorageUrl: string;
    user: string;
    key: string;
    cacheExpireAfterSeconds?: number;
}
/**
 * Simple Swift Object Storage client
 */
export declare class SwiftObjectStorage {
    private options;
    tokenCache: Cache;
    constructor(options: SwiftObjectStorageOptions);
    /**
     * Handles token authentication. Keeps the token data  cached so that
     * a new token doesn't have to be generated on every request
     */
    getTokenData(): Promise<TokenData>;
    /**
     * Returns a Request readable stream
     * @param container
     * @param filename
     */
    getRawGetStream(container: string, filename: string): Promise<Request>;
    getObjectWriteToStream(writeStream: WriteStream, container: string, filename: string, options?: GetObjectAsStreamOptions): Promise<{}>;
    /**
     * Returns a Request writeable stream
     * @param container
     * @param filename
     * @param options
     */
    getRawCreateStream(container: any, filename: string, options?: CreateObjectFromStreamOptions): Promise<Request>;
    createObjectFromStream(sourceStream: ReadStream, container: string, filename: string, options?: CreateObjectFromStreamOptions): Promise<{}>;
}


lib/SwiftObjectStorage.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = __importDefault(require("request"));
const node_cache_1 = __importDefault(require("node-cache"));
const pump_1 = __importDefault(require("pump"));
const SWIFT_TOKEN_KEY = "SWIFT_TOKEN_KEY";
const DEFAULT_EXP_SECONDS = 60 * 20;
/**
 * Simple Swift Object Storage client
 */
class SwiftObjectStorage {
    constructor(options) {
        this.options = options;
        this.tokenCache = new node_cache_1.default();
    }
    /**
     * Handles token authentication. Keeps the token data  cached so that
     * a new token doesn't have to be generated on every request
     */
    getTokenData() {
        return __awaiter(this, void 0, void 0, function* () {
            const tokenData = this.tokenCache.get(SWIFT_TOKEN_KEY);
            if (tokenData) {
                return Promise.resolve(tokenData);
            }
            return new Promise((resolve, reject) => {
                request_1.default.get({
                    url: this.options.objectStorageUrl,
                    headers: {
                        "X-Auth-Key": this.options.key,
                        "X-Auth-User": this.options.user
                    }
                }, (err, resp) => {
                    if (resp.statusCode !== 200) {
                        return reject(Error(`Error generating token: ${resp.statusCode} - ${resp.statusMessage}`));
                    }
                    const data = JSON.parse(resp.body);
                    const newTokenData = {
                        objectPath: data.storage.public,
                        authToken: resp.headers["x-auth-token"]
                    };
                    const expAfterSeconds = this.options.cacheExpireAfterSeconds || DEFAULT_EXP_SECONDS;
                    this.tokenCache.set(SWIFT_TOKEN_KEY, newTokenData, this.options.cacheExpireAfterSeconds);
                    return resolve(newTokenData);
                });
            });
        });
    }
    /**
     * Returns a Request readable stream
     * @param container
     * @param filename
     */
    getRawGetStream(container, filename) {
        return __awaiter(this, void 0, void 0, function* () {
            const tokenData = yield this.getTokenData();
            return Promise.resolve(request_1.default.get({
                url: `${tokenData.objectPath}/${container}/${filename}`,
                headers: { "X-Auth-Token": tokenData.authToken }
            }));
        });
    }
    getObjectWriteToStream(writeStream, container, filename, options = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            const sourceStream = yield this.getRawGetStream(container, filename);
            return new Promise((resolve, reject) => {
                sourceStream.on("response", resp => {
                    if (Math.floor(resp.statusCode / 100) !== 2) {
                        reject(Error(`Error generating token: ${resp.statusCode} - ${resp.statusMessage}`));
                        return;
                    }
                    if (options.onResponse) {
                        options.onResponse(resp);
                    }
                });
                pump_1.default(sourceStream, writeStream, err => {
                    if (err) {
                        reject(err);
                        return;
                    }
                    resolve();
                });
            });
        });
    }
    /**
     * Returns a Request writeable stream
     * @param container
     * @param filename
     * @param options
     */
    getRawCreateStream(container, filename, options = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            const tokenData = yield this.getTokenData();
            return Promise.resolve(request_1.default.put({
                url: `${tokenData.objectPath}/${container}/${filename}`,
                timeout: 1000 * 60 * 10,
                headers: Object.assign({}, options.headers, { "X-Delete-After": options.expiresAfterSeconds, "X-Auth-Token": tokenData.authToken })
            }));
        });
    }
    createObjectFromStream(sourceStream, container, filename, options = {}) {
        return new Promise((resolve, reject) => __awaiter(this, void 0, void 0, function* () {
            const putStream = yield this.getRawCreateStream(container, filename, options);
            putStream.on("response", resp => {
                if (Math.floor(resp.statusCode / 100) !== 2) {
                    reject(Error(`Error generating token: ${resp.statusCode} - ${resp.statusMessage}`));
                    return;
                }
                if (options.onResponse) {
                    options.onResponse(resp);
                }
            });
            pump_1.default(sourceStream, putStream, err => {
                if (err) {
                    reject(err);
                    return;
                }
                resolve();
            });
        }));
    }
}
exports.SwiftObjectStorage = SwiftObjectStorage;
//# sourceMappingURL=SwiftObjectStorage.js.map